misk-hibernate-testing/src/main/kotlin/misk/jdbc/AbstractScaleSafetyChecks.kt

```diff
@@ -0,0 +1,154 @@
+package misk.jdbc
+
+import com.squareup.moshi.Moshi
+import misk.hibernate.Transacter
+import misk.moshi.adapter
+import misk.okio.split
+import misk.vitess.StartDatabaseService
+import mu.KotlinLogging
+import java.sql.Connection
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okio.BufferedSource
+import okio.ByteString.Companion.encodeUtf8
+import java.io.EOFException
+import java.sql.Timestamp
+import javax.inject.Inject
+
+internal data class Instruction(val Opcode: String, val Input: Instruction?) {
+  val isScatter: Boolean
+    get() {
+      return Opcode == "SelectScatter" || Input?.isScatter ?: false
+    }
+}
+
+internal data class QueryPlan(
+  val Original: String,
+  val Instructions: Instruction,
+  val ExecCount: Int
+) {
+  val isScatter: Boolean
+    get() = Instructions.isScatter
+}
+
+internal data class Variables(
+  /** Queries processed by plan type */
+  val QueriesProcessed: Map<String, Int>
+)
+
+abstract class AbstractScaleSafetyChecks @Inject constructor(
+  open val config: DataSourceConfig,
+  open val okHttpClient: OkHttpClient,
+  open val moshi: Moshi,
+  open val transacter: Transacter
+) : DataSourceDecorator {
+  internal abstract fun connect(): Connection?
+
+  internal fun isDml(query: String): Boolean {
+    val first = query
+      .replace(COMMENT_PATTERN, "")
+      .trimStart()
+      .toLowerCase()
+      .takeWhile { !it.isWhitespace() }
+    return DML.contains(first)
+  }
+
+  /**
+   * Digs into the MySQL log to find the last executed DML statement that passed through Vitess.
+   */
+  internal fun extractLastDmlQuery(): String? {
```

Comment:
This should be in the Vitess scale checker right? I don't think we would use this in a mysql checker...